File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        ans = []
        n = len(nums)
        def traverse(li , seen ):
            nonlocal n , ans ,nums
            if len(li) == n :
                ans.append(li)
                return 
            for i in range(n):
                if i not in seen:
                    traverse(li+[nums[i]] , seen+[i])
        se = []
        traverse([] , se[:])
        return ans

File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    vector<vector<int>> ans;
    void traverse( vector<int> li , set<int> seen , vector<int> &nums){
        if (li.size() == nums.size()){
            ans.push_back(li);
            return ;
        }
        for (int i = 0 ; i<nums.size() ; i++){
            if (seen.count(i) == 0){
                seen.insert(i);
                li.push_back(nums[i]);
                traverse(li , seen , nums);
                li.pop_back();
                seen.erase(i);
            }
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        set<int> seen;
        traverse({} ,seen , nums );
        return ans;
    }
};